File: backend/Services/Analyzer/RuleBasedEmailAnalyzer.cs
using System.Text.RegularExpressions;
using CareerMail.Api.Models.Enums;
using CareerMail.Api.Services;

namespace CareerMail.Api.Services.Analyzer;

public class RuleBasedEmailAnalyzer : IEmailAnalyzer
{
    private readonly IRecruiterIntelligenceService _recruiterIntelligenceService;

    public RuleBasedEmailAnalyzer(IRecruiterIntelligenceService recruiterIntelligenceService)
    {
        _recruiterIntelligenceService = recruiterIntelligenceService;
    }

    private static readonly string[] JunkPatterns = new[]
    {
        "new jobs posted", "just posted a", "jobs matching your", "top jobs for you", "daily leetcoding",
        "visa application", "rewards", "security alert", "survey to tell us",
        "registration code", "confirm your identity", "create password", "get career-ready",
        "preparing for 2027", "latest update from targetjobs", "recommended jobs",
        "job recommendations", "newsletter", "weekly digest", "promotional", "voices you should know",
        "asda rewards", "marketing", "webinar", "podcast", "account verification",
        "password reset", "two-factor", "otp code", "one-time passcode", "adobe for students",
        "groww digest", "groww", "binance", "crypto", "train ticket", "uber", "autofill", "jobright",
        "driving licence", "driving license", "dvla", "barclaycard", "credit card", "klarna", "scentbird",
        "aspora", "real estate", "toro's express", "uber eats", "loan application", "mortgage"
    };

    private static readonly string[] OpportunityPatterns = new[]
    {
        "new opportunity", "job opportunity", "career opportunity", "exciting opportunity",
        "opportunity for you", "opportunity at", "opportunity with", "role opening",
        "position open", "job opening", "we are hiring", "is hiring", "hiring for",
        "thought you'd be a great fit", "thought you might be a fit", "thought you'd be interested",
        "thought of you for", "job match", "job alert", "recruiter reachout",
        "reaching out regarding a role", "reaching out regarding an opportunity",
        "wanted to reach out regarding", "found your profile", "impressed with your background",
        "open role", "new role for", "role opening at", "explore opportunities",
        "new career opportunity", "great opportunity", "exciting role", "position available"
    };

    private static readonly string[] ConfirmationPatterns = new[]
    {
        "received your application", "received your job application", "received your application for",
        "received an application from you", "we received your application", "we have received your application",
        "we've received your application", "we received your application for", "application has been received",
        "your application has been received", "application received", "application was received",
        "thank you for applying", "thank you for your application", "thanks for applying",
        "thanks for your application", "thanks for your job application", "thank you for submitting your application",
        "thanks for submitting your application", "your application was sent to", "your application to",
        "job application is confirmed", "application confirmation", "position confirmed", "position is confirmed",
        "successfully submitted your application", "thanks for your interest in", "thank you for your interest in",
        "indeed application:", "your application has been submitted", "confirming your application",
        "applied for the role", "applied for the position", "applied to", "applied for the job",
        "acknowledgement of your application", "application acknowledgement"
    };

    private static readonly string[] RejectionPatterns = new[]
    {
        "chosen to move forward with other candidates", "chosen to pursue other candidates",
        "we've chosen to move forward with other candidates", "we have chosen to move forward with other candidates",
        "decided not to proceed", "decided not to move forward", "decided not to progress",
        "decided to pursue other candidates", "decided to move forward with other candidates",
        "move forward with other candidates", "pursue other candidates",
        "not moving forward with your application", "not moving forward with your candidacy",
        "will not be moving forward with your application", "will not be moving forward",
        "will not be proceeding", "will not be progressing", "not be progressing your application",
        "not to progress your application", "not to progress to the next stage",
        "not to take your application further", "not selected for this position", "not selected for the role",
        "not selected for an interview", "not selected for this role", "unsuccessful on this occasion",
        "application has been unsuccessful", "application was unsuccessful", "we will not be able to offer you",
        "we are unable to offer you", "not able to offer you an interview", "unable to proceed with your application",
        "position has now been filled", "we have filled the position", "no longer considering your application"
    };

    private static readonly string[] OfferPatterns = new[]
    {
        "pleased to offer you", "delighted to offer you", "thrilled to offer you",
        "offer of employment", "job offer for", "formal job offer", "conditional job offer",
        "formal offer", "extending an offer", "extend an offer", "offer letter",
        "welcome to the team", "employment agreement", "contract of employment"
    };

    private static readonly string[] InterviewInvitationPatterns = new[]
    {
        "invite you to an interview", "invite you to interview", "invited to interview",
        "invitation to interview", "like to invite you to", "pleased to invite you to",
        "schedule an interview", "schedule a call", "schedule a phone screen",
        "schedule a screening call", "schedule your next round", "book your interview",
        "select a time for your interview", "interview availability", "availability for a quick call",
        "availability for an interview", "interview with", "round of interviews"
    };

    private static readonly string[] AssessmentPatterns = new[]
    {
        "hackerrank", "codility", "codesignal", "testgorilla", "hirevue", "pymetrics",
        "online assessment", "coding challenge", "technical assessment", "online test",
        "complete your assessment", "assessment deadline", "take-home test", "take home project"
    };

    public AnalysisResult Analyze(string? subject, string? body, string? sender, string? senderEmail)
    {
        var s = subject ?? "";
        var b = body ?? "";
        var snd = sender ?? "";
        var sndEmail = senderEmail ?? "";

        var combined = $"{s} {b}".ToLowerInvariant();

        // 1. Junk Check
        if (JunkPatterns.Any(p => combined.Contains(p)))
        {
            // If it matches junk pattern and neither strong application confirmation nor offer nor rejection
            var hasStrongConfirmation = ConfirmationPatterns.Any(c => combined.Contains(c));
            if (!hasStrongConfirmation && !OfferPatterns.Any(o => combined.Contains(o)))
            {
                return AnalysisResult.NonJob();
            }
        }

        // Extract company first
        var company = ExtractCompany(s, b, snd, sndEmail);
        var jobTitle = ExtractRole(s, b);

        var recruiterInfo = _recruiterIntelligenceService.ExtractIntelligence(s, b, snd, sndEmail, company);

        // 2. Offer Check
        if (OfferPatterns.Any(o => combined.Contains(o)))
        {
            return new AnalysisResult
            {
                JobRelated = true,
                Company = company,
                JobTitle = jobTitle,
                Status = ApplicationStatus.OFFER,
                Classification = EmailClassification.OFFER,
                RecruiterName = recruiterInfo.Name,
                RecruiterEmail = recruiterInfo.Email,
                RecruiterTitle = recruiterInfo.Title,
                RecruiterPhone = recruiterInfo.Phone,
                RecruiterLinkedin = recruiterInfo.Linkedin,
                RecruiterType = recruiterInfo.Type,
                ContactConfidence = recruiterInfo.Confidence,
                ContactExtractionSource = recruiterInfo.Source,
                TimelineNote = "Job Offer extended!",
                Confidence = 95
            };
        }

        // 3. Rejection Check
        if (RejectionPatterns.Any(r => combined.Contains(r)))
        {
            return new AnalysisResult
            {
                JobRelated = true,
                Company = company,
                JobTitle = jobTitle,
                Status = ApplicationStatus.REJECTED,
                Classification = EmailClassification.REJECTION,
                RecruiterName = recruiterInfo.Name,
                RecruiterEmail = recruiterInfo.Email,
                RecruiterTitle = recruiterInfo.Title,
                RecruiterPhone = recruiterInfo.Phone,
                RecruiterLinkedin = recruiterInfo.Linkedin,
                RecruiterType = recruiterInfo.Type,
                ContactConfidence = recruiterInfo.Confidence,
                ContactExtractionSource = recruiterInfo.Source,
                TimelineNote = "Application status update: Not selected",
                Confidence = 95
            };
        }

        // 4. Interview Check
        if (InterviewInvitationPatterns.Any(i => combined.Contains(i)))
        {
            var isFinal = combined.Contains("final round") || combined.Contains("final interview");
            var (interviewTime, meetingLink) = ExtractInterviewDetails(b);

            return new AnalysisResult
            {
                JobRelated = true,
                Company = company,
                JobTitle = jobTitle,
                Status = isFinal ? ApplicationStatus.FINAL_INTERVIEW : ApplicationStatus.INTERVIEW,
                Classification = EmailClassification.INTERVIEW_INVITATION,
                InterviewDateTime = interviewTime,
                InterviewLink = meetingLink,
                InterviewType = isFinal ? "Final Round Interview" : "Technical Interview",
                RecruiterName = recruiterInfo.Name,
                RecruiterEmail = recruiterInfo.Email,
                RecruiterTitle = recruiterInfo.Title,
                RecruiterPhone = recruiterInfo.Phone,
                RecruiterLinkedin = recruiterInfo.Linkedin,
                RecruiterType = recruiterInfo.Type,
                ContactConfidence = recruiterInfo.Confidence,
                ContactExtractionSource = recruiterInfo.Source,
                TimelineNote = "Interview invitation received",
                Confidence = 90
            };
        }

        // 5. Assessment Check
        if (AssessmentPatterns.Any(a => combined.Contains(a)))
        {
            var deadline = ExtractDeadline(b);
            return new AnalysisResult
            {
                JobRelated = true,
                Company = company,
                JobTitle = jobTitle,
                Status = ApplicationStatus.ASSESSMENT,
                Classification = EmailClassification.ASSESSMENT,
                Deadline = deadline,
                RecruiterName = recruiterInfo.Name,
                RecruiterEmail = recruiterInfo.Email,
                RecruiterTitle = recruiterInfo.Title,
                RecruiterPhone = recruiterInfo.Phone,
                RecruiterLinkedin = recruiterInfo.Linkedin,
                RecruiterType = recruiterInfo.Type,
                ContactConfidence = recruiterInfo.Confidence,
                ContactExtractionSource = recruiterInfo.Source,
                TimelineNote = "Online assessment / coding challenge received",
                Confidence = 90
            };
        }

        // 6. Application Confirmation Check
        if (ConfirmationPatterns.Any(c => combined.Contains(c)))
        {
            return new AnalysisResult
            {
                JobRelated = true,
                Company = company,
                JobTitle = jobTitle,
                Status = ApplicationStatus.APPLIED,
                Classification = EmailClassification.APPLICATION_RECEIVED,
                RecruiterName = recruiterInfo.Name,
                RecruiterEmail = recruiterInfo.Email,
                RecruiterTitle = recruiterInfo.Title,
                RecruiterPhone = recruiterInfo.Phone,
                RecruiterLinkedin = recruiterInfo.Linkedin,
                RecruiterType = recruiterInfo.Type,
                ContactConfidence = recruiterInfo.Confidence,
                ContactExtractionSource = recruiterInfo.Source,
                TimelineNote = "Application confirmed",
                Confidence = 85
            };
        }

        // 7. Opportunity / Recruiter Message Check
        if (OpportunityPatterns.Any(o => combined.Contains(o)) || recruiterInfo.Type == RecruiterType.HUMAN_RECRUITER)
        {
            return new AnalysisResult
            {
                JobRelated = true,
                Company = company,
                JobTitle = jobTitle,
                Status = ApplicationStatus.APPLIED,
                Classification = EmailClassification.NEW_OPPORTUNITY,
                RecruiterName = recruiterInfo.Name,
                RecruiterEmail = recruiterInfo.Email,
                RecruiterTitle = recruiterInfo.Title,
                RecruiterPhone = recruiterInfo.Phone,
                RecruiterLinkedin = recruiterInfo.Linkedin,
                RecruiterType = recruiterInfo.Type,
                ContactConfidence = recruiterInfo.Confidence,
                ContactExtractionSource = recruiterInfo.Source,
                TimelineNote = "Recruiter outreach / discovered opportunity lead",
                Confidence = 80
            };
        }

        // Not identified as job related
        return AnalysisResult.NonJob();
    }

    private string ExtractCompany(string subject, string body, string sender, string senderEmail)
    {
        // 1. Try Subject: "Application to [Company]" or "Thank you for applying to [Company]"
        var subjMatch = Regex.Match(subject, @"(?i)(?:at|to|with)\s+([A-Za-z0-9&.\- ]+?)(?:\s*(?:for|as|\-|\(|\||$))");
        if (subjMatch.Success)
        {
            var cand = subjMatch.Groups[1].Value.Trim();
            if (cand.Length >= 2 && cand.Length <= 35 && !cand.Contains("your application", StringComparison.OrdinalIgnoreCase))
            {
                return CleanCompany(cand);
            }
        }

        // 2. Try Sender Name if not generic
        if (!string.IsNullOrWhiteSpace(sender) && !sender.Equals("Recruiter", StringComparison.OrdinalIgnoreCase) && !sender.Contains("@"))
        {
            var cleaned = Regex.Replace(sender, @"(?i)(Careers|Recruiting|Team|Talent|Jobs|No-Reply|Notifications)", "").Trim();
            if (cleaned.Length >= 2 && cleaned.Length <= 35)
            {
                return CleanCompany(cleaned);
            }
        }

        // 3. Try Sender Email Domain
        if (!string.IsNullOrWhiteSpace(senderEmail) && senderEmail.Contains("@"))
        {
            var domain = senderEmail[(senderEmail.IndexOf("@") + 1)..];
            var parts = domain.Split('.');
            if (parts.Length > 0 && !parts[0].Equals("gmail", StringComparison.OrdinalIgnoreCase) && !parts[0].Equals("outlook", StringComparison.OrdinalIgnoreCase) && !parts[0].Equals("greenhouse", StringComparison.OrdinalIgnoreCase) && !parts[0].Equals("lever", StringComparison.OrdinalIgnoreCase))
            {
                var name = parts[0];
                return char.ToUpper(name[0]) + name[1..];
            }
        }

        return "Unknown Company";
    }

    private string CleanCompany(string comp)
    {
        var res = Regex.Replace(comp, @"(?i)\b(team|ltd|limited|inc|corp|group|careers|jobs|recruitment|technologies|software|solutions)\b", "").Trim();
        return string.IsNullOrWhiteSpace(res) ? comp : res;
    }

    private string ExtractRole(string subject, string body)
    {
        var match = Regex.Match(subject, @"(?i)(?:for|as a|position:|role:)\s+([A-Za-z0-9\s/\-]+?)(?:at|with|in|\||-|$)", RegexOptions.IgnoreCase);
        if (match.Success)
        {
            var extracted = match.Groups[1].Value.Trim();
            if (extracted.Length > 3 && extracted.Length < 50)
            {
                return extracted;
            }
        }
        return "Software Engineer";
    }

    private (DateTime?, string?) ExtractInterviewDetails(string body)
    {
        string? link = null;
        var linkMatch = Regex.Match(body, @"https?://(?:[a-zA-Z0-9_\-]+\.)?(?:meet\.google\.com|zoom\.us|teams\.microsoft\.com)/[a-zA-Z0-9_\-\?=/]+");
        if (linkMatch.Success)
        {
            link = linkMatch.Value;
        }

        // Default to in 3 days at 11 AM if not strictly parsed
        var interviewTime = DateTime.UtcNow.AddDays(3).Date.AddHours(11);
        return (interviewTime, link);
    }

    private DateOnly? ExtractDeadline(string body)
    {
        return DateOnly.FromDateTime(DateTime.UtcNow.AddDays(5));
    }
}

File: backend/app/services/analyzer/rule_based_email_analyzer.py
import re
from datetime import datetime, timedelta
from app.models.enums import EmailClassification, ApplicationStatus
from app.services.analyzer.analysis_result import EmailAnalysisResult
from app.services.recruiter_intelligence_service import RecruiterIntelligenceService


class RuleBasedEmailAnalyzer:
    def __init__(self, recruiter_service: RecruiterIntelligenceService | None = None):
        self.recruiter_service = recruiter_service or RecruiterIntelligenceService()

    def analyze(self, sender: str, sender_email: str, subject: str, body: str) -> EmailAnalysisResult:
        result = EmailAnalysisResult()
        clean_subj = subject.strip()
        clean_body = body.strip()
        full_text = f"{clean_subj}\n{clean_body}".lower()

        # 1. Quick Junk Filter
        if self._is_junk_or_spam(sender_email, full_text):
            result.is_job_related = False
            return result

        # 2. Extract Company & Role
        company = self._extract_company(sender, sender_email, clean_subj, clean_body)
        role = self._extract_role(clean_subj, clean_body)
        result.detected_company = company
        result.detected_role = role

        # 3. Recruiter Intelligence Extraction
        result.recruiter_info = self.recruiter_service.extract_recruiter_info(
            sender, sender_email, clean_subj, clean_body, company
        )

        # 4. Classification Rules in Priority Order:
        # A. Job Offer (Highest Priority)
        if self._is_offer(full_text):
            result.is_job_related = True
            result.classification = EmailClassification.OFFER
            result.suggested_status = ApplicationStatus.OFFER
            result.confidence_score = 0.95
            result.timeline_title = "Formal Job Offer Received!"
            result.timeline_description = f"Offer letter received from {company or 'employer'}"
            return result

        # B. Interview Invitation / Scheduled
        if self._is_interview(full_text):
            result.is_job_related = True
            result.classification = EmailClassification.INTERVIEW_INVITATION
            result.suggested_status = ApplicationStatus.INTERVIEW
            result.confidence_score = 0.92
            result.timeline_title = "Interview Scheduled"
            result.timeline_description = f"Interview with {company or 'team'}"
            result.meeting_link = self._extract_meeting_link(clean_body)
            result.interview_type = "Technical Interview" if "technical" in full_text or "coding" in full_text else "Recruiter Screen"
            result.interview_date = datetime.utcnow() + timedelta(days=2, hours=3)
            return result

        # C. Assessment / Test
        if self._is_assessment(full_text):
            result.is_job_related = True
            result.classification = EmailClassification.ASSESSMENT
            result.suggested_status = ApplicationStatus.ASSESSMENT
            result.confidence_score = 0.90
            result.timeline_title = "Assessment / Coding Test Received"
            result.timeline_description = f"Online assessment for {role or 'the role'} from {company or 'employer'}"
            return result

        # D. Rejection
        if self._is_rejection(full_text):
            result.is_job_related = True
            result.classification = EmailClassification.REJECTION
            result.suggested_status = ApplicationStatus.REJECTED
            result.confidence_score = 0.92
            result.timeline_title = "Application Status Update: Not Moving Forward"
            result.timeline_description = f"Decision received from {company or 'employer'}"
            return result

        # E. New Opportunity / Recruiter Reachout
        if self._is_opportunity(full_text):
            result.is_job_related = True
            result.classification = EmailClassification.NEW_OPPORTUNITY
            result.suggested_status = ApplicationStatus.APPLIED
            result.confidence_score = 0.88
            result.timeline_title = "Recruiter Opportunity Inbound"
            result.timeline_description = f"Recruiter reachout from {company or 'hiring team'}"
            return result

        # F. Application Confirmation / Submitted
        if self._is_confirmation(full_text):
            result.is_job_related = True
            result.classification = EmailClassification.APPLICATION_RECEIVED
            result.suggested_status = ApplicationStatus.APPLIED
            result.confidence_score = 0.85
            result.timeline_title = "Application Confirmed"
            result.timeline_description = f"Application received by {company or 'employer'}"
            return result

        # G. General Job Related
        if self._is_general_job_related(full_text):
            result.is_job_related = True
            result.classification = EmailClassification.OTHER_JOB_RELATED
            result.confidence_score = 0.65
            return result

        return result

    def _is_junk_or_spam(self, email: str, text: str) -> bool:
        junk_senders = ["marketing@", "promo@", "deals@", "newsletter@", "uber@", "ubereats@", "deliveroo-order@", "amazon-order@", "receipts@"]
        if any(j in email.lower() for j in junk_senders):
            return True
        spam_keywords = ["unsubscribe", "privacy policy update", "your order has shipped", "tracking number", "promo code", "discount code"]
        if sum(1 for kw in spam_keywords if kw in text) >= 2 and not any(kw in text for kw in ["interview", "application", "hiring", "job offer", "recruiter"]):
            return True
        return False

    def _is_offer(self, text: str) -> bool:
        patterns = [
            r"offer of employment",
            r"pleased to offer you",
            r"delighted to offer you",
            r"formal offer",
            r"written offer",
            r"congratulations on your offer",
            r"offer letter attached",
            r"job offer for",
            r"compensation package"
        ]
        return any(re.search(p, text, re.IGNORECASE) for p in patterns)

    def _is_interview(self, text: str) -> bool:
        patterns = [
            r"invitation to interview",
            r"interview invitation",
            r"invite you to an interview",
            r"schedule an interview",
            r"schedule a call",
            r"technical interview",
            r"system design interview",
            r"coding interview",
            r"chat with the team",
            r"speak with our hiring manager",
            r"meet\.google\.com",
            r"zoom\.us\/j\/",
            r"calendly\.com",
            r"next round of interviews"
        ]
        return any(re.search(p, text, re.IGNORECASE) for p in patterns)

    def _is_assessment(self, text: str) -> bool:
        patterns = [
            r"online assessment",
            r"coding assessment",
            r"hackerrank",
            r"codility",
            r"codesignal",
            r"take-home challenge",
            r"technical test",
            r"complete the test"
        ]
        return any(re.search(p, text, re.IGNORECASE) for p in patterns)

    def _is_rejection(self, text: str) -> bool:
        patterns = [
            r"unfortunately",
            r"not moving forward",
            r"pursue other candidates",
            r"decided to move forward with other",
            r"other applicants",
            r"will not be moving forward",
            r"not selected for an interview",
            r"not a fit at this time",
            r"wish you the best in your job search"
        ]
        return any(re.search(p, text, re.IGNORECASE) for p in patterns)

    def _is_opportunity(self, text: str) -> bool:
        patterns = [
            r"exciting (?:new )?opportunity",
            r"came across your profile",
            r"saw your background",
            r"thought you might be interested in",
            r"reaching out regarding a role",
            r"would you be open for a call",
            r"new opening for",
            r"potential fit for"
        ]
        return any(re.search(p, text, re.IGNORECASE) for p in patterns)

    def _is_confirmation(self, text: str) -> bool:
        patterns = [
            r"thank you for applying",
            r"thanks for applying",
            r"application received",
            r"application submitted",
            r"received your application",
            r"we have received your resume",
            r"application confirmation"
        ]
        return any(re.search(p, text, re.IGNORECASE) for p in patterns)

    def _is_general_job_related(self, text: str) -> bool:
        patterns = [r"recruiter", r"hiring manager", r"talent acquisition", r"job application", r"candidate status", r"interview process"]
        return any(re.search(p, text, re.IGNORECASE) for p in patterns)

    def _extract_company(self, sender: str, email: str, subject: str, body: str) -> str | None:
        # Check domain first
        domain_match = re.search(r"@([a-zA-Z0-9.-]+)\.([a-zA-Z]{2,})", email)
        if domain_match:
            dom = domain_match.group(1).lower()
            if dom not in ["gmail", "yahoo", "hotmail", "outlook", "icloud", "mail", "greenhouse", "lever", "workable", "ashbyhq"]:
                return dom.capitalize()

        # Check subject patterns: "at [Company]" or "[Company] - Application"
        at_match = re.search(r"\bat\s+([A-Z][a-zA-Z0-9\s&]+?)(?:\s+for|\s+-\s+|\s+\(|$|\.|\!)", subject)
        if at_match:
            cand = at_match.group(1).strip()
            if len(cand) > 1 and len(cand.split()) <= 4:
                return cand

        # Check sender name
        if sender and not any(k in sender.lower() for k in ["notification", "careers", "talent", "support", "no-reply"]):
            parts = sender.split()
            if len(parts) <= 3:
                return sender.strip()

        return None

    def _extract_role(self, subject: str, body: str) -> str | None:
        patterns = [
            r"(?:role|position|job|for)\s*:\s*([A-Za-z\s\/\-\#\.\+]+?)(?:\s+at|\s+with|\s+\-|\s+\(|$|\.|\!)",
            r"(?:for the)\s+([A-Za-z\s\/\-\#\.\+]+?)\s+(?:role|position)",
            r"(Software Engineer|Backend Developer|Frontend Developer|Full Stack Engineer|Data Scientist|Machine Learning Engineer|DevOps Engineer|Python Developer)"
        ]
        for p in patterns:
            m = re.search(p, subject, re.IGNORECASE)
            if m:
                cand = m.group(1).strip()
                if 3 <= len(cand) <= 50:
                    return cand

        return "Software Engineer"

    def _extract_meeting_link(self, text: str) -> str | None:
        m = re.search(r"(https:\/\/(?:meet\.google\.com\/[a-z]{3}-[a-z]{4}-[a-z]{3}|[a-zA-Z0-9.-]+\.zoom\.us\/j\/\d+|teams\.microsoft\.com\/l\/meetup-join\/[^\s\>]+))", text)
        return m.group(1) if m else None
